skyd/cache.go

```diff
@@ -0,0 +1,168 @@
+	psc.mu.Lock()
+	defer psc.mu.Unlock()
+	for _, s := range skylinks {
+		psc.skylinks[s] = struct{}{}
+	}
+}
+
+// Contains returns true when the given skylink is in the cache.
+func (psc *PinnedSkylinksCache) Contains(skylink string) bool {
+	psc.mu.Lock()
+	defer psc.mu.Unlock()
+	_, exists := psc.skylinks[skylink]
+	return exists
+}
+
+// Diff returns two lists of skylinks - the ones that are in the given list but
+// are not in the cache (missing) and the ones that are in the cache but are not
+// in the given list (removed).
+func (psc *PinnedSkylinksCache) Diff(sls []string) (unknown []string, missing []string) {
+	psc.mu.Lock()
+	defer psc.mu.Unlock()
+	removedMap := make(map[string]struct{}, len(psc.skylinks))
+	for sl := range psc.skylinks {
+		removedMap[sl] = struct{}{}
+	}
+	for _, sl := range sls {
+		// Remove this skylink from the removedMap, because it has not been
+		// removed.
+		delete(removedMap, sl)
+		// If it's not in the cache - add it to the added list.
+		_, exists := psc.skylinks[sl]
+		if !exists {
+			unknown = append(unknown, sl)
+		}
+	}
+	// Transform the removed map into a list.
+	for sl := range removedMap {
+		missing = append(missing, sl)
+	}
+	return
+}
+
+// Rebuild rebuilds the cache of skylinks pinned by the local skyd. The
+// rebuilding happens in a goroutine, allowing the method to return a channel
+// on which the caller can either wait or select. The caller can check whether
+// the rebuild was successful by calling Error().
+func (psc *PinnedSkylinksCache) Rebuild(skydClient NodeSkydClient) RebuildCacheResult {
+	psc.mu.Lock()
+	defer psc.mu.Unlock()
+	if !psc.isRebuildInProgress() {
+		psc.result = &RebuildCacheResult{
+			Ch:        make(chan struct{}),
+			ExternErr: nil,
+		}
+		// Kick off the actual rebuild in a separate goroutine.
+		go psc.threadedRebuild(skydClient)
+	}
+	return *psc.result
+}
+
+// Remove registers the given skylinks in the cache.
+func (psc *PinnedSkylinksCache) Remove(skylinks ...string) {
+	psc.mu.Lock()
+	defer psc.mu.Unlock()
+	for _, s := range skylinks {
+		delete(psc.skylinks, s)
+	}
+}
+
+// isRebuildInProgress returns true if a cache rebuild is in progress.
+// Calling this method assumes that caller is holding a lock on the cache.
+func (psc *PinnedSkylinksCache) isRebuildInProgress() bool {
+	return psc.result != nil
+}
+
+// threadedRebuild performs the actual cache rebuild process. It reports any
+// errors by setting the psc.err variable and it always closes the rebuildCh on
+// exit.
+func (psc *PinnedSkylinksCache) threadedRebuild(skydClient NodeSkydClient) {
+	var err error
+	// Ensure that we properly wrap up the rebuild process.
+	defer func() {
+		psc.mu.Lock()
+		// Update the result.
+		psc.result.ExternErr = err
+		close(psc.result.Ch)
+		// Mark the rebuild as done.
+		psc.result = nil
+		psc.mu.Unlock()
+	}()
+
+	// Walk the entire Skynet folder and scan all files we find for skylinks.
+	dirsToWalk := []skymodules.SiaPath{skymodules.SkynetFolder}
```

Comment:
Follow Up: this is implemented in skyd now. You can call the client method `SkynetSkylinksGet()` to dedup this code.

Reply:
Already on the list: https://github.com/SkynetLabs/pinner/issues/3